lib/ddtrace/transport/http/client.rb

```diff
@@ -0,0 +1,94 @@
+require 'ddtrace/encoding'
+require 'ddtrace/version'
+
+require 'ddtrace/transport/traces'
+require 'ddtrace/transport/http/api'
+require 'ddtrace/transport/http/api_map'
+require 'ddtrace/transport/http/endpoint'
+
+require 'ddtrace/transport/http/compatibility'
+
+module Datadog
+  module Transport
+    module HTTP
+      # Routes, encodes, and sends tracer data to the trace agent via HTTP.
+      class Client
+        include Compatibility
+
+        attr_reader \
+          :service,
+          :apis,
+          :active_api,
+          :encoder,
+          :headers
+
+        DEFAULT_APIS = APIMap[
+          V4 = 'v0.4'.freeze => API.new(
+            Transport::Traces::Parcel => TracesEndpoint.new('/v0.4/traces'.freeze, Encoding::MsgpackEncoder)
+          ),
+          V3 = 'v0.3'.freeze => API.new(
+            Transport::Traces::Parcel => TracesEndpoint.new('/v0.3/traces'.freeze, Encoding::MsgpackEncoder)
+          ),
+          V2 = 'v0.2'.freeze => API.new(
+            Transport::Traces::Parcel => TracesEndpoint.new('/v0.2/traces'.freeze, Encoding::JSONEncoder)
+          )
+        ].with_fallbacks(V4 => V3, V3 => V2).freeze
+
+        def initialize(service, options = {})
+          @service = service
+          @apis = options[:apis] || DEFAULT_APIS
+
+          # Select active API
+          @active_api = apis[options.fetch(:api_version, V3)]
+          raise UnknownApiVersion if active_api.nil?
+
+          # Select encoder
+          @encoder = options[:encoder]
```

Comment:
don't we want to grab this from the `@active_api`?

Reply:
I'll have to look at this again.

Reply:
From what I can tell, I think this suggestion would be fine. I don't remember why I exactly wrote it this way: maybe because I was reworking what was in the old transport, maybe I wanted some kind of override (much of the code seems to allow for encoder to be overridden), but I don't remember exactly. Either way, it should be safe to set it to whatever the `active_api` is.

Reply:
Okay, actually can't do this, because API doesn't define encoding, the endpoint does (which makes sense, since different endpoints can have encodings.)

I think it'd be better to consider if we need to force encoders on endpoints from the client (which is essentially the purpose it serves.) If not, we can just eliminate this option entirely from `Client`, have it be handled only in `ParcelEndpoint`, which would probably simplify the code without sacrificing much.

Reply:
This is obsolete. Encoding has been incorporated into the `Datadog::Transport::HTTP::Traces::API::Endpoint` which defines how requests are prepared.